### ISSUE TITLE
Improve accessibility and usability of multi-value fields

### ISSUE THREAD

**Issue by justcolin** [projecthydra/sufia#2021]
### Descriptive summary

In the course of performing user testing with Sufia, @bnhowell found that the "Add" and "Remove" buttons were causing a number of problems for users. They had trouble figuring out and remembering what they did. We found that repositioning them and adding the name of the field to the buttons helped users a great deal.

The "Remove" buttons in the proposed example below are visible text instead of just an "X" icon to make them easier for people using voice control software to operate. We also styled the buttons like links because the big green and red buttons were drawing too much attention for a secondary feature.

As a compromise to make sure the "Remove" buttons do not get too wide on mobile devices, I propose not including the name of the field in the "Remove" buttons, though this is up for discussion.

The same update would also benefit Curation Concerns for the same reasons.

### Example

![screen shot 2016-05-13 at 11 26 43 am copy](https://cloud.githubusercontent.com/assets/11234702/15253651/426d6a5c-1901-11e6-9db2-66668f0ae8df.png)

### Related work

#1612

**Comment by justcolin** [comment]
@projecthydra/sufia-ui-ux-advisors

**Comment by newmanld** [comment]
Seems like user testing has validated this approach.  Curious if the entire form is longer or shorter.

**Comment by newmanld** [comment]
In regard to accessibility - any plans for accessibility scan?  We were dinged by an accessibility scan (and I suspect different scans highlight different things) for issues we would not have thought about.  One was that when we had a list of files and download buttons in line with each file, the download buttons (at least the title text behind the buttons) needed to be distinct.  This might apply to the Add or Remove links?  https://github.com/uclibs/scholar_uc/issues/521

**Comment by mtribone** [comment]
Yeah similarly to what @newmanld has described above, I was wondering @justcolin if we could use Bootstrap's `sr only` to keep the context, Remove Creator, for users of assistive devices?

**Comment by mtribone** [comment]
also tagging @jonearley about `sr-only`

**Comment by heidisb** [comment]
Big improvement. All are logical changes made to improve both usability and accessibility. Much clearer and cleaner.

**Comment by jonearley** [comment]
@mtribone Appending the field label to the remove button wouldn't add much value in this case since the remove action is referring only to the adjacent field where the label is for the entire multi-field. I think it does make sense to append the label on the "add another..." button since that action is referring to the entire multi-field.

Maybe `sr-only` could read as "remove adjacent field" to help clarify. Thoughts?

**Comment by justcolin** [comment]
It is a couple of `em`s longer, but not enough to be a real difference.

**Comment by mtribone** [comment]
@jonearley I guess I misread the initial thread that adding the label to the button was helpful, but because of mobile was to be removed. The text "remove adjacent field" _might_ be helpful. Hopefully the inputs and controls are grouped properly in HTML. I should go check on that...

**Comment by justcolin** [comment]
@newmanld: These changes only make the fields a couple of `em`'s longer so the length isn't a problem.

As for accessibility scans, I ran some scans as well as hand testing with some screen readers a little while ago (that's where are lot of these issues are originating from). Once we get through the list of issues from that scan I was planning on running another scan and some more usage tests to see what else we can do.

As for repeated text: automated scans flag that because it is often—but not always—a bad thing and is easy for a computer to check for. In this case it is pretty much unavoidable and doesn't cause any real problems. It is those "Click here!" links that are really bad. The important thing here is that it is clear in context what each button does from the label.

**Comment by justcolin** [comment]
re:@mtribone: I've looked into adding screen-reader-only text there and contextually it is pretty obvious what is going on, but I was thinking of adding "previous field" after "Remove" for clarity (since adjacent is a visual cue). However, I'll ask a real screen reader user that we know what he thinks.

**Comment by justcolin** [comment]
@mtribone and @jonearley: Our screen reader specialist said that they would prefer it if each button had text which read as "Remove previous Author". "Remove" should be visible, the rest should be `sr-only`.

**Comment by mjgiarlo** [comment]
Tagging CurationConcerns folks so they can see this: @jpstroop @escowles @tpendragon

**Comment by escowles** [comment]
👍  this looks like a great improvement

**Comment by bnhowell** [comment]
Let's test it and see.

**Comment by justcolin** [comment]
@bnhowell: Test which part?

**Comment by tpendragon** [comment]
:+1: Pretty down with anything that improves accessibility, and think we should move it to CC. Thank you so much for looking into this.

**Comment by justcolin** [comment]
@jonearley is currently writing a PR which add this change to the Hydra gem that CC and Sufia get these fields.

**Comment by bnhowell** [comment]
@justcolin We should test the Add FieldName and Remove controls with a blind user as well as a low vision user (more difficult to find a low vision user)

**Comment by jcoyne** [comment]
The widget as draw appears to  allow the user to remove every entry.  I'm pretty sure we don't want to allow that.

**Comment by justcolin** [comment]
@jcoyne: Users want to be able to remove any field, so if the last field is no longer applicable they can remove it. However, if you remove fields until only one is left the remove button will disappear so there is always at least field.

**Comment by jonearley** [comment]
Here is a UI preview from the PR I'm working on.

- When an add or remove action is made I have focus set to the last field.
- The list of fields has aria-live=polite set so when that region changes a screen reader will report when idle (e.g. a new field was added or removed). https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions
- The remove button for screen readers reads as "Remove previous Title field".

Your thoughts and suggestions are welcome!

![multi-field animated preview with keyboard control](http://g.recordit.co/tjQIADKZj5.gif)

**Comment by mjgiarlo** [comment]
@jonearley Looks great! Thank you for working on this.

**Comment by simosacchi** [comment]
+1 looking really good!

**Comment by bnhowell** [comment]
Looks so good Jon! The ARIA Live Regions are a great solution too!
What are the small grey icons on the right side of the entry fields? What
do they activate?

On Tue, May 17, 2016 at 11:49 AM, Simone Sacchi <notifications@github.com>
wrote:

> +1 looking really good!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/projecthydra/sufia/issues/2021#issuecomment-219762045>
>



-- 
Ben Howell | Accessibility & User Experience Specialist
Design & Discovery | LIT | University of Michigan Library
bnhowell@gmail.com

**Comment by jonearley** [comment]
@bnhowell You can ignore those small grey icons, that's my lastpass Chrome extension.

**Comment by justcolin** [comment]
@jonearley: Do you think you'll be able to finish your PR for this next week?

**Comment by atz** [comment]
Not that we have dealt with i18n in any serious way yet, but note that "previous" would only make sense in `ltr` languages.  I think that's acceptable for foreseeable releases.

**Comment by justcolin** [comment]
Good point! That text should be set in the locale files anyways so it could become part of any translation work. I haven't tested it, but is Sufia/CC designed to be easily restyled for right-to-left languages?

**Comment by justcolin** [comment]
@jonearley I believe we were talking about moving the "remove" and "add" text into the `locales` files, correct?

**Comment by atz** [comment]
In the example animated image, note that the Creator's `+` button also reads "Add another Title".

**Comment by mjgiarlo** [comment]
Re: rtl languages... good question. I am not sure!

**Comment by mjgiarlo** [comment]
Oops, sorry for temporarily closing the issue. Misfire.

**Comment by jonearley** [comment]
@atz Good catch, but that's fixed in the most recent PR.

**Comment by jonearley** [comment]
This PR has been merged into hydra-editor. Thanks for the help @jcoyne: https://github.com/projecthydra/hydra-editor/pull/113

@justcolin I agree about moving the text into the `locals` files. I'll start on a second PR for that.

**Comment by mjgiarlo** [comment]
@jonearley will that second PR close this issue? (Getting a handle on what's remaining for Sufia 7.0.0.)

**Comment by jonearley** [comment]
@mjgiarlo I'd say it's OK to close this issue now without that second PR.

**Comment by mjgiarlo** [comment]
OK, thanks, @jonearley
